[FX] Normalize torch. namespace op arguments

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54236 [FX] Normalize Python operators to `torch.` ops when appropriate
* **#53832 [FX] Normalize torch. namespace op arguments**
* #53831 [FX] Experimental type annotation pass using Python signatures
* #53830 Function to retrieve inspect.Signature instances for PyTorch ops

Extend the experimental `normalize` pass to also cover operators on the `torch.` namespace. This utilizes the new `operator_schemas.get_signature_for_torch_op` introduced in #53830

Differential Revision: [D26982801](https://our.internmc.facebook.com/intern/diff/D26982801)

Commit:
[FX] Normalize torch. namespace ops

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize torch. namespace ops"


Differential Revision: [D26982801](https://our.internmc.facebook.com/intern/diff/D26982801)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize torch. namespace ops"


Differential Revision: [D26982801](https://our.internmc.facebook.com/intern/diff/D26982801)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize torch. namespace ops"


Differential Revision: [D26982801](https://our.internmc.facebook.com/intern/diff/D26982801)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize torch. namespace ops"


Differential Revision: [D26982801](https://our.internmc.facebook.com/intern/diff/D26982801)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize torch. namespace ops"


Differential Revision: [D26982801](https://our.internmc.facebook.com/intern/diff/D26982801)

[ghstack-poisoned]

Chillee commented on torch/fx/operator_schemas.py:
```diff
@@ -44,7 +44,12 @@ def _torchscript_schema_to_signature(ts_schema : torch._C.FunctionSchema) -> ins
         arg_type = _torchscript_type_to_python_type(arg.type)
         default = arg.default_value if arg.has_default_value() else inspect.Parameter.empty
-        name = arg.name
+        # TODO: Figure out if this is safe. It seems like when generating the type signatures for
```
Can you explain a bit more about what the problem is here? I'm not sure what the context is.

jamesr66a replied:
The schemas use `self` as the name for the first argument, e.g. https://github.com/pytorch/pytorch/blob/407d60ee910175cecc3a95617175105779fbbcbd/aten/src/ATen/native/native_functions.yaml#L424

But `python_torch_functions.cpp` is generated s.t. the first argument is named `input` (e.g. https://github.com/pytorch/pytorch/blob/407d60ee910175cecc3a95617175105779fbbcbd/tools/autograd/templates/python_torch_functions.cpp#L566 and see the generated code in your built repo)

I can't figure out where exactly this switch is happening so the TODO here

Chillee commented on torch/fx/experimental/normalize.py:
```diff
@@ -44,27 +47,40 @@ def call_function(self, target : Target, args : Tuple[Argument, ...], kwargs : D
+            assert callable(target)
+            torch_op_schemas = get_signature_for_torch_op(target)
+            if torch_op_schemas:
+                for candidate_signature in torch_op_schemas:
```
Perhaps add a comment here about what this is doing (i.e. `iterates through possible overloads to determine which one matches`).

Btw, is it possible that there are multiple overloads that this can bind to?

jamesr66a replied:
> Btw, is it possible that there are multiple overloads that this can bind to?

Yes, but AFAIK the signatures should be in the same order as in overload resolution, which uses the first one that matches the given arguments

Chillee reviewed (approved):
LGTM - some clarity questions but otherwise good (assuming that `get_signature_for_torch_ops` is good).

Commit:
Update on "[FX] Normalize torch. namespace ops"


Extend the experimental `normalize` pass to also cover operators on the `torch.` namespace. This utilizes the new `operator_schemas.get_signature_for_torch_op` introduced in #53830

Differential Revision: [D26982801](https://our.internmc.facebook.com/intern/diff/D26982801)

[ghstack-poisoned]
